### DISCUSSION TITLE
`Fastlane Produce` for Play Store?

### DISCUSSION THREAD

**Discussion by sooxt98** [fastlane/fastlane#16858] [Ideas] [answered]
fastlane to send the first Android publication

**Comment by joshdholtz** [comment, marked as answer]
@sooxt98 I really wish this was possible but sadly its not 😔 The Google Play API SDK does not give us this ability. I’ve seen some people do some web scraping / headless browser scripting to accomplish this but this isn’t something we can do and maintain in _fastlane_.

**Reply by androidseb** [reply]
My understanding is that the current implementation of `fastlane produce` for App Store Connect is using private browser APIs to perform its operations (looking at the [App Store Connect API](https://developer.apple.com/documentation/appstoreconnectapi) provided by Apple, it doesn't seem to be possible to create new apps), maybe I got this wrong though?

Isn't the Google Play Console website just a web app sending REST API calls to interact with the Google Play Store?

Instead of doing web scraping, could it be possible to use the REST API used by the Google Play Console directly from fastlane's ruby code? Or is that against Google's EULA?

@joshdholtz Do you have any thoughts about this? I'm assuming you already considered this so I'm probably missing something here.

**Reply by joshdholtz** [reply]
@androidseb You are correct! `produce` is currently using the private/legacy APIs for creating developer portal apps but the web version of App Store Connect API for App Store apps. 

The Google Play Console is similar in the way it sends REST API calls but the Google auth is so much more complex than the Apple ID 😔 I'm also not sure if its against Google's EULA.

We currently make use of the Ruby SDK that Google provides for all other API calls. Introducing something against this REST API would present users with all whole other type of auth which be a little confusing and so complex to handle with 2FA and what not 🤷‍♂️ 

I'm totally fine if somebody wants to create a plugin that does this but its probably not something that we will be adding into fastlane until the official Google SDK supports this 😇

**Reply by max-ott** [reply]
Google obfuscates most of it's web stuff for spam-reasons and similar concerns. I highly doubt that we will be able to work on this and that it is appreciated by Google. Most Android/`supply`-related stuff is based on the Google Play Publishing API (https://developers.google.com/android-publisher) Check out the functionality there. This is what is possible and within reason.

**Reply by androidseb** [reply]
Thank you both for your replies.

@max-ott I took a quick glance at the network tab of my Google Play Console, and I could see that the requests results are indeed not JSON, they are total gibberish, I had not thought of that. I guess this means they want to explicitly discourage such use of the Google Play Console, so most likely not a good idea to go with what I suggested then :-(

@joshdholtz I see how handling 2 separate authentication models will increase complexity, but my first thoughts were that it's not like Fastlane doesn't already have something similar for iOS: because of Apple 2FA, you guys had to add `FASTLANE_SESSION` and `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD`.

I was thinking maybe we could make a tool to create and store a Google web cookie session, an equivalent to `fastlane spaceauth -u ...` but now that I see that:
* Google has an **official** way with service accounts
* The Google Play Console browser REST API is obfuscated

I can totally understand your position on the matter, it makes sense.

In the early days of Fastlane, when Apple did not even have any APIs for their Store, I remember you guys doing magic with web-crawling, but I also remember it was fragile and a lot of work to maintain (I'm very grateful for the Fastlane project for that by the way).

Hopefully Google will add official support for this in the future, thanks for your time :-)

**Reply by joshdholtz** [reply]
@androidseb Its not _really_ the handling of two auths that is difficult just the two auth types that we have to use with Google that makes it difficult. If you want to take a stab and mark a PR or a plugin that would be great! I’d be happy to take a look and test it out when it comes time but I don’t think this is something that we can put into our development queue at the moment 😔

Unless things change with the Google SDK 😉
